### PR TITLE
Scikit learn like cubical interface

### PR DESCRIPTION
**Pull request by VincentRouvreau**

### CODE REVIEW

File: src/python/gudhi/sklearn/cubical_persistence.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            homology_coeff_field=self.homology_coeff_field, min_persistence=self.min_persistence
+        )
+        return cubical_complex.persistence_intervals_in_dimension(self.only_this_dim)
+
+    def transform(self, X, Y=None):
+        """
+        Compute all the cubical complexes and their associated persistence diagrams.
+
+        Parameters:
+            X (list of list of double OR list of numpy.ndarray): List of cells filtration values that can be flatten if
```
Comment (Hind-M) [inline src/python/gudhi/sklearn/cubical_persistence.py:93]: Should the list be flattened before putting it as input or a list of list should be working ?
```suggestion
            X (list of list of double OR list of numpy.ndarray): List of cells filtration values that can (or should ?) be flattened if
```

File: src/python/doc/cubical_complex_user.rst [hunk 2]
```diff
@@ -163,4 +168,82 @@ Tutorial
...
+    X, y = fetch_openml("mnist_784", version=1, return_X_y=True, as_frame=False)
+    
+    # Target is: "is an eight ?"
+    y = (y == "8") * 1
+    print("There are", np.sum(y), "eights out of", len(y), "numbers.")
+    
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)
+    pipe = Pipeline(
+        [
+            ("cub_pers", CubicalPersistence(only_this_dim=0, dimensions=[28, 28], n_jobs=-2)),
```
Comment (Hind-M) [inline src/python/doc/cubical_complex_user.rst:214]: Should we add a note explaining that using `CubicalPersistence` followed by `DimensionSelector` is equivalent to using `CubicalPersistence` with `only_this_dim` ? Or maybe add a commented line of code as a "second" example to depict this option ?

File: src/python/gudhi/sklearn/cubical_persistence.py (hunk 1 above)
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py]: Thanks, `it should be flatten if...`. On f55ae925

File: src/python/gudhi/sklearn/cubical_persistence.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                to be reshaped (cf. :func:`~gudhi.sklearn.cubical_persistence.CubicalPersistence.transform`)
+            max_persistence_dimension (int): The returned persistence diagrams maximal dimension. Default value is `0`.
+                Ignored if `only_this_dim` is set.
+            only_this_dim (int): The returned persistence diagrams dimension. If `only_this_dim` is set,
+                `max_persistence_dimension` will be ignored. 
+                Short circuit the use of :class:`~gudhi.sklearn.post_processing.DimensionSelector` when only one
+                dimension matters.
+            homology_coeff_field (int): The homology coefficient field. Must be a prime number. Default value is 11.
+            min_persistence (float): The minimum persistence value to take into account (strictly greater than
+                `min_persistence`). Default value is `0.0`. Sets `min_persistence` to `-1.0` to see all values.
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py]: Fixed on f55ae925

File: src/python/doc/cubical_complex_user.rst (hunk 2 above)
Comment (VincentRouvreau) [reply src/python/doc/cubical_complex_user.rst]: I had a comment to show to deal when `persistence_dimension` is an array on f55ae925. Example still with `persistence_dimension` is an integer.

File: src/python/gudhi/sklearn/cubical_persistence.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        Compute all the cubical complexes and their associated persistence diagrams.
+
+        Parameters:
+            X (list of list of double OR list of numpy.ndarray): List of cells filtration values that should be flatten
+                if `dimensions` is set in the constructor, or already with the correct shape in a numpy.ndarray (and
+                `dimensions` must not be set).
+
+        Returns:
+            list of pairs or list of list of pairs:
+            Persistence diagrams in the format:
```
Comment (mglisse) [inline src/python/gudhi/sklearn/cubical_persistence.py:94]: I didn't try, would removing ":" avoid the bold font for this whole line?

File: src/python/doc/cubical_complex_sklearn_itf_ref.rst [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    X, y = fetch_openml("mnist_784", version=1, return_X_y=True, as_frame=False)
+    
+    # Target is: "is an eight ?"
+    y = (y == "8") * 1
+    print("There are", np.sum(y), "eights out of", len(y), "numbers.")
+    
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)
+    pipe = Pipeline(
+        [
+            ("cub_pers", CubicalPersistence(persistence_dimension=0, dimensions=[28, 28], n_jobs=-2)),
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/doc/cubical_complex_sklearn_itf_ref.rst:57]: @mglisse Both work (and same timings). Which one do you prefer to promote ?

File: src/python/doc/cubical_complex_sklearn_itf_ref.rst [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+Cubical complex persistence scikit-learn like interface
+#######################################################
+
+.. list-table::
+   :widths: 40 30 30
+   :header-rows: 0
+
+   * - :Since: GUDHI 3.5.0
+     - :License: MIT
+     - :Requires: `Scikit-learn <installation.html#scikit-learn>`_
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/doc/cubical_complex_sklearn_itf_ref.rst:14]: Done on b6f4e23 and open issue #579 as some other tables are ugly

File: src/python/gudhi/sklearn/cubical_persistence.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            homology_coeff_field=self.homology_coeff_field, min_persistence=self.min_persistence
+        )
+        return cubical_complex.persistence_intervals_in_dimension(self.persistence_dimension)
+
+    def transform(self, X, Y=None):
+        """
+        Compute all the cubical complexes and their associated persistence diagrams.
+
+        Parameters:
+            X (list of list of double OR list of numpy.ndarray): List of cells filtration values that should be flatten
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py:88]: Done on c34cd02 for `src/python/gudhi/sklearn/cubical_persistence.py` and `src/python/gudhi/representations/preprocessing.py`.

File: src/python/test/test_sklearn_cubical_persistence.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+from gudhi.sklearn.cubical_persistence import CubicalPersistence
+import numpy as np
+from sklearn import datasets
+
+CUBICAL_PERSISTENCE_H0_IMG0 = np.array([[0.0, 6.0], [0.0, 8.0], [0.0, np.inf]])
+
+
+def test_simple_constructor_from_top_cells():
+    cells = datasets.load_digits().images[0]
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/test/test_sklearn_cubical_persistence.py:19]: When you install sklearn, it installs a subpackage `datasets` that contains `data/digits.csv.gz`. This is this file that is read when you call `datasets.load_digits()`, so no, there is no network need to call this method.

File: src/python/gudhi/sklearn/cubical_persistence.py (hunk 4 above)
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py:94]: Use [sphinx docstring](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html) for line not to be bold. Done on 36457ad

File: src/python/gudhi/sklearn/cubical_persistence.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        homology_coeff_field=11,
+        min_persistence=0.0,
+        n_jobs=None,
+    ):
+        """
+        Constructor for the CubicalPersistence class.
+
+        Parameters:
+            dimensions (list of int): A list of number of top dimensional cells if cells filtration values will require
+                to be reshaped (cf. :func:`~gudhi.sklearn.cubical_persistence.CubicalPersistence.transform`)
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py:46]: I wanted to stick to the [cubical class constructor documentation](https://gudhi.inria.fr/python/latest/cubical_complex_ref.html#gudhi.CubicalComplex.__init__), but you are right, maybe something like [newshape](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html#numpy.reshape) would be more appropriate.
I tested it on this [gist skcub.py](https://gist.github.com/VincentRouvreau/db57e57ab12ba66dbc7a68b2242b6494).
I will add an unitary test.

File: src/python/gudhi/representations/preprocessing.py [hunk 10]
```diff
@@ -363,3 +364,51 @@ def __call__(self, diag):
...
+        """
+        return self
+
+    def transform(self, X, Y=None):
+        """
+        Select persistence diagrams from its dimension.
+
+        Parameters:
+            X (list of list of pairs): List of list of persistence pairs, i.e.
+                `[[array( Hi(X0) ), array( Hj(X0) ), ...], [array( Hi(X1) ), array( Hj(X1) ), ...], ...]`
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/representations/preprocessing.py:406]: @mglisse Do you mean that, as there is no pair in python, it is called tuple ? (list of list of tuples)

File: src/python/doc/cubical_complex_sklearn_itf_ref.rst [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+   returns its persistence diagrams
+#. :class:`~gudhi.representations.DiagramSelector` that removes non-finite persistence diagrams values
+#. :class:`~gudhi.representations.PersistenceImage` that builds the persistence images from persistence diagrams
+#. `SVC <https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html>`_ which is a scikit-learn support
+   vector classifier.
+
+This ML pipeline is trained to detect if the hand written digit is an '8' or not, thanks to the fact that an '8' has
+two holes in :math:`\mathbf{H}_1`, or, like in this example, three connected components in :math:`\mathbf{H}_0`.
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/doc/cubical_complex_sklearn_itf_ref.rst:32]: I like the 'example driven development' approach, and I think that an example of how to use gudhi in a sklearn pipeline is what the users are waiting for. But, sure, CubicalPersistence is only one line of code in this example...

File: src/python/gudhi/sklearn/cubical_persistence.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/gudhi/sklearn/cubical_persistence.py]: I don't think the current version is clear enough. When I read that it changes the shape without any precision, I expect it to do the same thing as `numpy.reshape`, but it doesn't. Maybe it should. Or if we want to say that it is passed as parameter `dimensions` to `CubicalComplex.__init__`, we need to be explicit about it... but I can't seem to find where we document the meaning of that parameter in CubicalComplex :confused: There would also be the possibility of taking a function as argument that is applied to each point, so the user could pass `lambda a:a.reshape(m,n,p)`, but that's probably overkill.

File: src/python/gudhi/representations/preprocessing.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/representations/preprocessing.py:406]: Ok that's it, tuple is the right type. Done on 3bfc0665

File: src/python/doc/cubical_complex_sklearn_itf_ref.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/doc/cubical_complex_sklearn_itf_ref.rst]: Use numpy reshape on [8d7244b](https://github.com/GUDHI/gudhi-devel/pull/499/commits/8d7244b510eee0d7927c521117198ef286028acf). It removes any ambiguity.

File: src/python/gudhi/sklearn/cubical_persistence.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py]: Use numpy reshape on [8d7244b](https://github.com/GUDHI/gudhi-devel/pull/499/commits/8d7244b510eee0d7927c521117198ef286028acf). It removes any ambiguity.

File: src/python/gudhi/sklearn/cubical_persistence.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        min_persistence=0.0,
+        n_jobs=None,
+    ):
+        """
+        Constructor for the CubicalPersistence class.
+
+        Parameters:
+            newshape (list of int): A list of number of top dimensional cells if cells filtration values will require
+                to be reshaped (cf. :func:`~gudhi.sklearn.cubical_persistence.CubicalPersistence.transform`)
+            persistence_dimension (int or list of int): The returned persistence diagrams dimension(s).
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py]: Renamed argument as homology_dimensions on [8d7244b](https://github.com/GUDHI/gudhi-devel/pull/499/commits/8d7244b510eee0d7927c521117198ef286028acf) to be consistent with #411

File: src/python/gudhi/sklearn/cubical_persistence.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+class CubicalPersistence(BaseEstimator, TransformerMixin):
+    """
+    This is a class for computing the persistence diagrams from a cubical complex.
+    """
+
+    def __init__(
+        self,
+        newshape=None,
+        persistence_dimension=-1,
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py]: @mglisse Behind the hood, CubicalPersistence is using [persistence_intervals_in_dimension](https://gudhi.inria.fr/python/latest/cubical_complex_ref.html#gudhi.CubicalComplex.persistence_intervals_in_dimension) which returns an empty list when the dimension given as an argument has no persistence values (which is the case for `-1`).
Setting `persistence_dimension` to `-1` will always return empty diagrams.
What do you think would be a *"good"* default value ? 
* `0` to return H0 by default ?
* Or `-1` to return empty diagrams by default ?

I will adapt the documentation regarding our decision.

File: src/python/doc/cubical_complex_sklearn_itf_ref.rst [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+Cubical complex persistence scikit-learn like interface example
+---------------------------------------------------------------
+
+In this example, hand written digits are used as an input.
+a TDA scikit-learn pipeline is constructed and is composed of:
+
+#. :class:`~gudhi.sklearn.cubical_persistence.CubicalPersistence` that builds a cubical complex from the inputs and
+   returns its persistence diagrams
+#. :class:`~gudhi.representations.DiagramSelector` that removes non-finite persistence diagrams values
+#. :class:`~gudhi.representations.PersistenceImage` that builds the persistence images from persistence diagrams
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/doc/cubical_complex_sklearn_itf_ref.rst]: Good catch ! I modified it on [106e48d](https://github.com/GUDHI/gudhi-devel/pull/499/commits/106e48d65bf86aa75ec7bc317b3d008da4c23c5d) and the links are ok now.

File: src/python/gudhi/sklearn/cubical_persistence.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/gudhi/sklearn/cubical_persistence.py]: How about **not** having a default value?
The list `[0, 1, dim-1]` could also make sense I guess.

File: src/python/gudhi/sklearn/cubical_persistence.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+        """Compute all the cubical complexes and their associated persistence diagrams.
+
+        :param X: List of cells filtration values (`numpy.reshape(X, newshape, order='C'` if `newshape` is set with a tuple of ints).
+        :type X: list of list of float OR list of numpy.ndarray
+
+        :return: Persistence diagrams in the format:
+
+              - If `homology_dimensions` was set to `n`: `[array( Hn(X[0]) ), array( Hn(X[1]) ), ...]` 
+              - If `homology_dimensions` was set to `[i, j]`: `[[array( Hi(X[0]) ), array( Hj(X[0]) )], [array( Hi(X[1]) ), array( Hj(X[1]) )], ...]`
+        :rtype: list of tuple or list of list of tuple
```
Comment (mglisse) [inline src/python/gudhi/sklearn/cubical_persistence.py:96]: The type and explanation above don't match. The first case looks like a list of (*,2) arrays, while the second looks like a list of list of (*,2) arrays.

File: src/python/gudhi/sklearn/cubical_persistence.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py]: Ok, I left it without default value, just like CubicalLayer, on e4384664

File: src/python/gudhi/sklearn/cubical_persistence.py (hunk 15 above)
Comment (VincentRouvreau) [reply src/python/gudhi/sklearn/cubical_persistence.py]: I have checked on numpy documentation (like [this one](https://numpy.org/doc/stable/reference/generated/numpy.tensordot.html) i.e.) and I like the way they are writting it (`(2,) array_like`), so I stand with this syntax. Modified to `list of (,2) array_like or list of list of (,2) array_like` on [69198e9](https://github.com/GUDHI/gudhi-devel/pull/499/commits/69198e9a00648aa5f8f38e1cef2c7bd6b7299dbb)

**Review by mglisse — APPROVED** [review]
